// eslint-plugin-flowtype-errors/dist/collect.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _toArray2 = require('babel-runtime/helpers/toArray');

var _toArray3 = _interopRequireDefault(_toArray2);

exports.default = Collect;

var _path = require('path');

var _path2 = _interopRequireDefault(_path);

var _child_process = require('child_process');

var _child_process2 = _interopRequireDefault(_child_process);

var _slash = require('slash');

var _slash2 = _interopRequireDefault(_slash);

var _shelljs = require('shelljs');

var _shelljs2 = _interopRequireDefault(_shelljs);

var _filter = require('./filter');

var _filter2 = _interopRequireDefault(_filter);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var flowBin = void 0; /**
                       * Run Flow and collect errors in JSON format
                       *
                       * Reference the following links for possible bug fixes and optimizations
                       * https://github.com/facebook/nuclide/blob/master/pkg/nuclide-flow-rpc/lib/FlowRoot.js
                       * https://github.com/ptmt/tryflow/blob/gh-pages/js/worker.js
                       */


try {
  if (!process.env.FLOW_BIN) {
    flowBin = require('flow-bin'); // eslint-disable-line global-require
  }
} catch (e) {
  /* eslint-disable */
  console.log();
  console.log('Oops! Something went wrong! :(');
  console.log();
  console.log('eslint-plugin-flowtype-errors could not find the package "flow-bin". This can happen for a couple different reasons.');
  console.log();
  console.log('1. If ESLint is installed globally, then make sure "flow-bin" is also installed globally.');
  console.log();
  console.log('2. If ESLint is installed locally, then it\'s likely that "flow-bin" is not installed correctly. Try reinstalling by running the following:');
  console.log();
  console.log('  npm i -D flow-bin@latest');
  console.log();
  process.exit(1);
  /* eslint-enable */
}

// Adapted from https://github.com/facebook/flow/blob/master/tsrc/flowResult.js

function mainLocOfError(error) {
  var operation = error.operation,
      message = error.message;

  return operation && operation.loc || message[0].loc;
}

/**
 * Wrap critical Flow exception into default Error json format
 */
function fatalError(stderr) {
  return {
    errors: [{
      message: [{
        path: '',
        code: 0,
        line: 0,
        start: 0,
        descr: stderr
      }]
    }]
  };
}

function _formatMessage(message, messages, root, path) {
  switch (message.type) {
    case 'Comment':
      return '' + message.descr;
    case 'Blame':
      {
        var see = message.path !== '' ? ' See ' + (path === message.path ? 'line ' + message.line : '.' + (0, _slash2.default)(message.path.replace(root, '')) + ':' + message.line) + '.' : '';
        return '\'' + message.descr + '\'.' + see;
      }
    default:
      return '\'' + message.descr + '\'.';
  }
}

function getFlowBin() {
  return process.env.FLOW_BIN || flowBin;
}

function executeFlow(stdin, root, filepath) {
  var stdout = void 0;

  switch (stdin && root && filepath && stdin !== '') {
    case true:
      stdout = new _shelljs2.default.ShellString(stdin).exec(['' + getFlowBin(), 'check-contents --json --root', root + ' ' + filepath].join(' '), { silent: true });
      break;
    default:
      stdout = _child_process2.default.spawnSync(getFlowBin(), ['--json']).stdout;
  }

  //
  // This serves as a temporary HACK to prevent 32 bit OS's from failing. Flow does not
  // support 32 bit OS's at the moment.
  // This pretends as if there are now flow errors
  //
  // Ideally, there would be a preinstall npm event to check if the user is on a 32 bit OS
  //

  if (!stdout) {
    return true;
  }

  var stringifiedStdout = stdout.toString();
  var parsedJSONArray = void 0;

  try {
    parsedJSONArray = JSON.parse(stringifiedStdout);
  } catch (e) {
    parsedJSONArray = fatalError(stringifiedStdout);
  }

  var fullFilepath = _path2.default.resolve(root, filepath);

  // Loop through errors in the file
  var output = parsedJSONArray.errors
  // Temporarily hide the 'inconsistent use of library definitions' issue
  .filter(function (error) {
    var mainLoc = mainLocOfError(error);
    var mainFile = mainLoc && mainLoc.source;
    return mainFile && error.message[0].descr && !error.message[0].descr.includes('inconsistent use of') && _path2.default.resolve(root, mainFile) === fullFilepath;
  }).map(function (error) {
    var message = error.message,
        operation = error.operation;

    var _concat = [].concat(operation || [], message),
        _concat2 = (0, _toArray3.default)(_concat),
        firstMessage = _concat2[0],
        remainingMessages = _concat2.slice(1);

    var entireMessage = firstMessage.descr + ': ' + remainingMessages.reduce(function (previousMessage, currentMessage, index, messages) {
      return previousMessage + ' ' + _formatMessage(currentMessage, messages, root, firstMessage.path);
    }, '');

    return (0, _extends3.default)({}, process.env.DEBUG_FLOWTYPE_ERRRORS === 'true' ? parsedJSONArray : {}, {
      message: entireMessage.replace(/\.$/, ''),
      path: firstMessage.path,
      start: firstMessage.loc.start.line,
      end: firstMessage.loc.end.line,
      loc: firstMessage.loc
    });
  });

  return output.length ? (0, _filter2.default)(output) : true;
}

function Collect(stdin, root, filepath) {
  return executeFlow(stdin, root, filepath, {});
}
module.exports = exports['default'];
